cli/flox-rust-sdk/src/models/environment/core_environment.rs

```diff
@@ -515,6 +570,66 @@ impl CoreEnvironment<ReadOnly> {
         })
     }
 
+    fn upgrade_with_catalog_client(
+        &mut self,
+        flox: &Flox,
+        client: &impl ClientTrait,
+        groups_or_iids: &[String],
+        manifest: &TypedManifestCatalog,
+    ) -> Result<UpgradeResult, CoreEnvironmentError> {
+        let mut existing_lockfile = 'lockfile: {
+            let Ok(lockfile_path) = CanonicalPath::new(self.lockfile_path()) else {
+                break 'lockfile None;
+            };
+            let lockfile = LockedManifest::read_from_file(&lockfile_path)
+                .map_err(CoreEnvironmentError::LockedManifest)?;
+            match lockfile {
+                LockedManifest::Catalog(lockfile) => Some(lockfile),
+                _ => {
+                    warn!(
+                        "Found version 1 manifest, but lockfile doesn't match: Ignoring lockfile."
+                    );
+                    None
+                },
+            }
+        };
+
+        if let Some(ref mut lockfile) = existing_lockfile {
+            let locked_packages = std::mem::take(&mut lockfile.packages);
+
+            lockfile.packages = locked_packages
+                .into_iter()
+                .filter(|pkg| {
+                    !(groups_or_iids.contains(&pkg.group)
+                        || groups_or_iids.contains(&pkg.install_id))
+                })
+                .collect();
+        }
+
+        let upgraded =
+            LockedManifestCatalog::lock_manifest(manifest, existing_lockfile.as_ref(), client)
+                .block_on()
+                .map_err(CoreEnvironmentError::LockedManifest)?;
+
+        let store_path =
+            self.transact_with_lockfile_contents(serde_json::json!(upgraded).to_string(), flox)?;
+
+        let result = UpgradeResult {
+            packages: upgraded
+                .packages
+                .clone()
+                .into_iter()
+                .filter(|pkg| {
+                    groups_or_iids.contains(&pkg.group) || groups_or_iids.contains(&pkg.install_id)
+                })
+                .map(|pkg| pkg.install_id)
+                .collect(),
+            store_path: Some(store_path),
+        };
+
+        Ok(result)
+    }
```

Comment:
blocking: needs tests that groups and iids are handled correctly

Reply:
Added a case for iid == group